### ISSUE TITLE
Bad bundle filename generated on Windows

### ISSUE THREAD

**Issue by seanmiddleditch** [webpack/webpack-dev-middleware#270]
* Operating System: Windows 10 x64
* Node Version: 8.4.0
* NPM Version: 5.3.0
* webpack version: 4.0.1
* webpack-dev-middleware Version: 3.0.0

- [ ] This is a **feature** request
- [x] This is a **bug**

### Code

webpack.config.js snippet:

```js
    output: {
        filename: 'bundle.js',
        path: path.join(__dirname, 'dist'),
        publicPath: '/js/',
        libraryTarget: 'window'
    },
```

server express snippet:

```js
        const webpack = require('webpack')
        const webpackDevMiddleware = require('webpack-dev-middleware')
        const config = require('webpack.config.js')
        const compiler = webpack(config)
        app.use(webpackDevMiddleware(compiler, {
            publicPath: config.output.publicPath
        }))
```

### Expected Behavior

webpack-dev-middleware is able to serve the generated bundles correctly on Windows.

Given an `output.path` of `c:\workspace\dist` and a filename of `bundle.js` the generated filename from `getFilenameFromUrl` should be: `c:\workspace\dist\bundle.js`

### Actual Behavior

Fails to find the generated bundle due to a malformed Windows path when joining the bundle name and the output path, as the bundle path generated is in the form: `/c://\workspace\dist/bundle.js`.

This appears to be because of code like https://github.com/webpack/webpack-dev-middleware/blob/78e98794e5fff0728c1d33657cddbcc192a6ccb7/lib/util.js#L94 that uses `url-join` instead of Node's `path` module for forming file paths, plus the bit at https://github.com/webpack/webpack-dev-middleware/blob/78e98794e5fff0728c1d33657cddbcc192a6ccb7/lib/util.js#L96 that slaps `/` at the start of a path instead of using `path.resolve`. Those malformed URIs are then passed into the filesystem which naturally fails to find the file.

Replacing that chunk of code at the end of `getFilenameFromUrl` with the simpler `return path.resolve(outputPath || '', filename)` fixes the problem for me, though I'd hardly be surprised if that broke other use cases, so the proper fix is perhaps more intricate.

I note that some other use cases in the code like https://github.com/webpack/webpack-dev-middleware/blob/78e98794e5fff0728c1d33657cddbcc192a6ccb7/lib/middleware.js#L49 are making the same mistake of using `url-join` instead of `path`, but they aren't affecting my use case so I haven't needed to touch them.

I wouldn't be surprised if the current wdm code breaks in some rather unlikely setups with POSIX paths, too. As similar as POSIX paths and URIs are, they don't have exactly identical rules and it's a mistake to treat them as interchangeable.

### For Bugs; How can we reproduce the behavior?

Run a webpack-dev-middleware example on Windows.

### For Features; What is the motivation and/or use-case for the feature?

Developing on Windows workstations.

**Comment by shellscape** [comment]
Thanks for the issue 🍺 

`url-join` is only being used in situations where the middleware is serving the file based on the url requested, so I'm not sure I agree with your assessment there that it's incorrect. The leading slashes when referencing the filesystem are an issue though.

We'd welcome a pull request!

**Comment by shellscape** [comment]
This would be a great first contribution for someone running on Windows!

**Comment by KulakovSerg** [comment]
Is it a bug?
I use something like this everywhere in config paths:
`output: {
    path: path.resolve(process.cwd(), 'public/js'),`
and have no any problems...

**Comment by shellscape** [comment]
@KulakovSerg I can't verify because I don't use windows (that referenced commit above was a guess, but I have to wait for appveyor to get setup to verify). We know that some of the tests are failing from a user who ran the tests on his windows machine. The guess it that it's a problem with the transition from a url path to a filesystem path.

**Comment by shellscape** [comment]
@seanmiddleditch @asolopovas can you guys try using this fork to see if it resolves the issue on your end as well? https://github.com/montogeek/webpack-dev-middleware

**Comment by seanmiddleditch** [comment]
@shellscape - your change still doesn't work unfortunately. By the file that new `path.resolve(filename)` is called, the value in `filename` is already malformed, and the resolve just makes it worse. :)

I started on a CL to fix everything up to resolve paths correctly but ran into problems with the filesystem abstraction in place, and with getting test cases to resolve nicely. e.g., `MemoryFileSystem` doesn't provide a version of `resolve`. Ultimately the simplest fix I found was to replace the `uri.startsWith('/')` to `pathabs(uri)`. Basically the same as the other PR from monto, but using the `path-is-absolute` module that's already imported by that module.

The path generated is still weird (it's being formatted as a URI, so it turns into `c://workspace\foo` instead of `c:\workspace\foo`) but it at least works (Node is either not using Windows' Unicode UNC support or Node is doing a great job translating paths under the hood, it seems).

PR incoming.

**Comment by shellscape** [comment]
@seanmiddleditch did you try @montogeek's fork here https://github.com/montogeek/webpack-dev-middleware? he's been running tests locally on windows and says that resolves the failing tests on his end.

**Comment by seanmiddleditch** [comment]
@shellscape - commented on PR 275. It works, but I'd personally prefer my change since it allows tests for both POSIX and NT paths to execute on the Travis CI infrastructure.

A "better" change still would probably need to abstract out the path generation to match that expected by the underlying `context.fs` filesystem wrapper. That is, the `getFilename` function probably shouldn't depend directly on the host environment's native behavior; it should only use behavior specified by `context.fs`, since the path string is only ever used with `context.fs` functions. Unfortunately, that object is (usually) provided from Webpack's compiler object so any change here is scoped beyond just the dev middleware.

A compromise change would be to parameterize `getFilenameFromUrl` over an optional path resolver function, which could default to `path.resolve`. That would allow tests to explicitly provide either `path.posix.resolve` or `path.win32.resolve` as appropriate while still falling back to non-pure behavior outside of tests. The problem I ran into prototyping that approach was that the degree of changes I had to make to the test suite itself was beyond what I felt comfortable making as a newb to the codebase.
